packages/back-end/src/services/experiments.ts

```diff
@@ -3199,3 +3225,70 @@ export async function computeResultsStatus({
     },
   };
 }
+
+export async function validateExperimentData(
+  context: ReqContext,
+  data: Partial<ExperimentInterfaceStringDates>,
+  res: Response
+): Promise<
+  { metricIds: string[]; datasource: DataSourceInterface | null } | undefined
+> {
+  let datasource: DataSourceInterface | null = null;
+  if (data.datasource) {
+    datasource = await getDataSourceById(context, data.datasource);
+    if (!datasource) {
+      res.status(403).json({
+        status: 403,
+        message: "Invalid datasource: " + data.datasource,
+      });
+      return;
+    }
+  }
+
+  // Validate that specified metrics exist and belong to the organization
+  const metricIds = getAllMetricIdsFromExperiment(data);
+  if (metricIds.length) {
+    const map = await getMetricMap(context);
+    for (let i = 0; i < metricIds.length; i++) {
+      const metric = map.get(metricIds[i]);
+      if (metric) {
+        // Make sure it is tied to the same datasource as the experiment
+        if (data.datasource && metric.datasource !== data.datasource) {
+          res.status(400).json({
+            status: 400,
+            message:
+              "Metrics must be tied to the same datasource as the experiment: " +
+              metricIds[i],
+          });
+          return;
+        }
+      } else {
+        // check to see if this metric is actually a metric group
+        const metricGroup = await context.models.metricGroups.getById(
+          metricIds[i]
+        );
+        if (metricGroup) {
+          // Make sure it is tied to the same datasource as the experiment
+          if (data.datasource && metricGroup.datasource !== data.datasource) {
+            res.status(400).json({
+              status: 400,
+              message:
+                "Metric group must be tied to the same datasource as the experiment: " +
+                metricIds[i],
+            });
+            return;
+          }
+        } else {
+          // new metric that's not recognized...
+          res.status(403).json({
+            status: 403,
+            message: "Unknown metric: " + metricIds[i],
+          });
+          return;
+        }
+      }
+    }
+  }
```

Comment:
Diff this with the controller change to make sure nothing was missed accidentally